lib/index.js

```diff
@@ -20,13 +20,48 @@ function throttlingPlugin (octokit) {
     triggersNotificationLimiter: new Bottleneck({
       maxConcurrent: 1,
       minTime: 3000
-    })
-  }
+    }),
+    retryLimiter: new Bottleneck()
+  }, octokitOptions.throttle)
+
+  const events = {}
+  const emitter = new Bottleneck.Events(events)
+  events.on('abuse-limit', state.onAbuseLimit)
+  events.on('rate-limit', state.onRateLimit)
+  events.on('error', e => console.warn('Error in throttling-plugin limit handler', e))
+
+  state.retryLimiter.on('failed', async function (error, info) {
+    const options = info.args[info.args.length - 1]
+    const retryCount = ~~options.request.retryCount
+    options.request.retryCount = retryCount
+
+    if (error.status === 403 && /\babuse\b/i.test(error.message)) {
+      // The user has hit the abuse rate limit.
+      // https://developer.github.com/v3/#abuse-rate-limits
+
+      // The Retry-After header can sometimes be blank when hitting an abuse limit,
+      // but is always present after 2-3s, so make sure to set `retryAfter` to at least 5s by default.
+      const retryAfter = Math.max(~~error.headers['retry-after'], state.minimumAbuseRetryAfter)
+      const wantRetry = await emitter.trigger('abuse-limit', retryAfter, options)
+
+      if (wantRetry) {
+        options.request.retryCount++
+        return retryAfter * state.retryAfterBaseValue
+      }
+    } else if (error.status === 403 && error.headers['x-ratelimit-remaining'] === '0') {
+      // The user has used all their allowed calls for the current time period
+      // https://developer.github.com/v3/#rate-limiting
+
+      const rateLimitReset = new Date(~~error.headers['x-ratelimit-reset'] * 1000).getTime()
+      const retryAfter = Math.max(Math.ceil((rateLimitReset - Date.now()) / 1000), 0)
+      const wantRetry = await emitter.trigger('rate-limit', retryAfter, options)
 
-  octokit.throttle = {
-    _options: (options = {}) => Object.assign(state, options)
-  }
-  const emitter = new Bottleneck.Events(octokit.throttle)
+      if (wantRetry) {
+        options.request.retryCount++
+        return retryAfter * state.retryAfterBaseValue
+      }
+    }
+  })
```

Comment:
This is only a question of style, we can leave it as is :)

I usually try to avoid `else` by returning early. I’d write the above code like this

```js
if (error.status !== 403) {
  return
}

let wantRetry

if (/\babuse\b/i.test(error.message)) {
  // The user has hit the abuse rate limit.
  // https://developer.github.com/v3/#abuse-rate-limits

  // The Retry-After header can sometimes be blank when hitting an abuse limit,
  // but is always present after 2-3s, so make sure to set `retryAfter` to at least 5s by default.
  const retryAfter = Math.max(~~error.headers['retry-after'], state.minimumAbuseRetryAfter)
  wantRetry = await emitter.trigger('abuse-limit', retryAfter, options)
} 

if (error.headers['x-ratelimit-remaining'] === '0') {
  // The user has used all their allowed calls for the current time period
  // https://developer.github.com/v3/#rate-limiting

  const rateLimitReset = new Date(~~error.headers['x-ratelimit-reset'] * 1000).getTime()
  const retryAfter = Math.max(Math.ceil((rateLimitReset - Date.now()) / 1000), 0)
  wantRetry = await emitter.trigger('rate-limit', retryAfter, options)
}

if (wantRetry) {
  options.request.retryCount++
  return retryAfter * state.retryAfterBaseValue
}
```

Reply:
I changed it to something in between your suggestion and a more functional style